iree/compiler/Dialect/Flow/Transforms/FusionOfTensorOps.cpp

```diff
@@ -65,60 +91,20 @@ struct FusionOfTensorOpsPass
                       consumer->operand_end());
       if (operands.size() >= kIreeMaxOperandCount) return false;
 
-      bool isBroadcast = false;
-      if (auto genericOp = dyn_cast<linalg::GenericOp>(producer)) {
-        // Detect op that only broadcast input as fusing them makes the new
-        // op cheaper.
-        if (genericOp.getNumParallelLoops() == genericOp.getNumLoops() &&
-            isa<linalg::YieldOp>(genericOp.getBody()->front())) {
-          for (OpOperand *opOperand : genericOp.getInputOperands()) {
-            AffineMap indexingMap = genericOp.getTiedIndexingMap(opOperand);
-            if (indexingMap.isProjectedPermutation() &&
-                indexingMap.getNumDims() != indexingMap.getNumResults()) {
-              isBroadcast = true;
-              break;
-            }
-          }
-        }
-      }
-      // Only fuse if it has a single linalg generic user. It is a
-      // simplistic heuristic to avoid duplicating ops that may be
-      // expensive.
-      // TODO: Add a cost model to allow ops to be duplicated.
-      bool hasI1ReturnType =
-          llvm::any_of(producer->getResultTypes(), [](Type t) {
-            if (t.isInteger(1)) return true;
-            if (auto shapedType = t.dyn_cast<ShapedType>()) {
-              if (shapedType.getElementType().isInteger(1)) return true;
-            }
-            return false;
-          });
-      if (!isBroadcast && !isa<arith::ConstantOp>(producer) &&
-          !hasI1ReturnType &&
-          !llvm::hasSingleElement(producerResult.getUsers())) {
-        return false;
-      }
-      return llvm::all_of(producerResult.getUsers(), [](Operation *user) {
-        return isa<linalg::GenericOp>(user);
-      });
+      return areFusableOps(context, producer, consumer);
     };
     linalg::populateElementwiseOpsFusionPatterns(fusionPatterns, controlFn);
 
-    // Simple heuristic to decide if reshaope should be folded in the linalg.
-    // If the source of the reshape is a linalg op fold to potentially allow the
-    // two linalg ops to be fused. Otherwise leave it to avoid adding dimensions
-    // to the consumer linalg op.
-    linalg::ControlFusionFn foldReshapeBetweenLinalgFn =
+    // Always fold reshape by expansion.
+    linalg::ControlElementwiseOpsFusionFn fuseByExpansionControlFn =
         [](const OpResult &producer, const OpOperand &consumer) {
-          auto collapseOp = producer.getDefiningOp<tensor::CollapseShapeOp>();
-          if (collapseOp) {
-            return collapseOp.src().getDefiningOp<LinalgOp>() != nullptr;
+          // Do not fuse producer generic op if it has more than one user.
```

Comment:
It would be nice to have a TODO comment if we want to relax the condition in future.